### PR TITLE
add: pref package

### PR DESCRIPTION
**Pull request by Inkeliz**

### CODE REVIEW

**Review by whereswaldon — CHANGES_REQUESTED** [review]
Wow, this is really solid work! I do have some requests:

- I'd prefer at least a little documentation for the internal packages
- I think that we should provide a mechanism to differentiate between "this API isn't available" and the maximum return value of a function. Preferably an error associated with the function call.
- I believe that the java code and JNI could be simplified through the use of static methods. See my inline comments for details.
- The `go:generate` comments currently only work on Windows, as they use backslash-delimited paths. While I understand that you primarily develop on windows, I'd prefer for these to use UNIX paths so that I can regenerate the JAR files myself for future maintenance. I did some research on whether there's a way to use OS-agnostic paths in `go:generate`, but didn't come up with anything. If you know of a way, please tell me!

Overall though, this is an excellent addition to the Gio ecosystem, and I very much look forward to merging it. Thank you for submitting it.

**Comment by Inkeliz** [comment]
> I'd prefer at least a little documentation for the internal packages

> I think that we should provide a mechanism to differentiate between "this API isn't available" and the maximum return value of a function. Preferably an error associated with the function call.

I think it's better to export the internal package. So, call `battery.Level()` instead of `pref.Battery().Level`. Then, each function can return a `error` (`battery.Level() (uint8, error)` ). Or, maybe put everything together into a single `pref` package?

> * The `go:generate` comments currently only work on Windows, as they use backslash-delimited paths. While I understand that you primarily develop on windows, I'd prefer for these to use UNIX paths so that I can regenerate the JAR files myself for future maintenance. I did some research on whether there's a way to use OS-agnostic paths in `go:generate`, but didn't come up with anything. If you know of a way, please tell me!

Using the "linux path" works on Windows too. In the same way that `$TEMP` works on Windows (which uses `%TEMP%`). I didn't notice that before, but it's a easy fix. :)

File: pref/pref.go
Comment (Inkeliz) [reply pref/pref.go:38]: Some APIs (Javascript) doesn't expose that information. I don't know about Android, since most Android devices uses batteries, so maybe it could be hardcode to `true`?! 🤔

The Windows API have that information, so it could be added.

**Comment by whereswaldon** [comment]
Just popping in to say that I've had very little free time in my evenings this week, so I haven't been able to review your changes yet. I'm hoping to do it tomorrow night.

File: pref/internal/locale/locale_android.java
Comment (Inkeliz) [reply pref/internal/locale/locale_android.java]: Seems that the jni library doesn't have `CallStaticBooleanMethod`/`CallStaticIntMethod`. only `CallBooleanMethod`. The String (getLanguage) might work, but other options may not.

Comment (whereswaldon) [reply pref/internal/locale/locale_android.java]: It is missing the boolean one, but it has `CallStaticIntMethod`: https://pkg.go.dev/git.wow.st/gmp/jni#CallStaticIntMethod

File: pref/internal/battery/battery_js.go
Comment (Inkeliz) [reply pref/internal/battery/battery_js.go]: If the device doesn't use a battery, it's always fully charged, so it returns 100.

However, I include a new error, which may happen when batteries don't power the device. Some API doesn't provide such information, so it will return 100 without error. The JS API returns "100" when the machine doesn't use a battery, and there's no way to distinguish that (not officially).

File: pref/internal/locale/locale_android.java
Comment (whereswaldon) [reply pref/internal/locale/locale_android.java]: I have a PR up for implementing boolean support: https://git.wow.st/gmp/jni/pulls/5

We can leave the booleans as non-static until that's merged (it won't block acceptance of this PR), but I'd like to change the others to their non-static forms.

File: pref/battery/battery_windows.go
Comment (whereswaldon) [inline pref/battery/battery_windows.go:43]:
```suggestion
	if resp.BatteryFlag &= _BatteryFlagNoSystemBattery > 0 {
```

File: pref/battery/battery_windows.go
Comment (whereswaldon) [inline pref/battery/battery_windows.go:27]:
```suggestion
	if resp.BatteryLifePercent == _BatteryLifePercentUnknown || resp.BatteryFlag &= _BatteryFlagNoSystemBattery > 0 {
```
Given that this is a single-bit flag, I think this is the way to check it. I assume that the other bits will all be zero anyway if this is the case, but I don't think it will hurt to treat it like a flag.

File: pref/battery/battery_windows.go
Comment (whereswaldon) [inline pref/battery/battery_windows.go:56]:
```suggestion
	if resp.BatteryFlag &= _BatteryFlagNoSystemBattery > 0 {
```

File: pref/battery/battery_windows.go
Comment (whereswaldon) [inline pref/battery/battery_windows.go:52]: I think this comparison is correct, as the Unknown status option is more of a sentinel value than a flag (it sets all of the bits)

File: pref/battery/battery_windows.go
Comment (whereswaldon) [inline pref/battery/battery_windows.go:60]:
```suggestion
	return resp.BatteryFlag &= _BatteryFlagCharging > 0, nil
```

File: pref/doc.go
Comment (whereswaldon) [inline pref/doc.go:1]: Did you intend to write package-level documentation here?

**Review by whereswaldon — CHANGES_REQUESTED** [review]
This is looking really good! I think the API is good as-is, but the JNI code can still be simplified and there are some bitwise comparisons I'd like to look at in the windows code. Additionally, it looks like you created a file to write the package-level documentation in, but the docs aren't there. Did they get missed in your commits?

Thanks again for working on this! It'll be really useful for the community.

File: pref/internal/locale/locale_android.java
Comment (Inkeliz) [reply pref/internal/locale/locale_android.java]: I change to Int, and then compare `i >= 1`. 😁

**Comment by Inkeliz** [comment]
I change the Boolean to Int, and it's now using Static. To avoid code-duplication I make one `internal/xjni`, since now everything is using `Int`/`String`.

I add some description on README.md, with some small example. The `locale` now supports Linux (tested on Fedora) and I add the `*_unsupported.go` for any other OS.

I fix the bitwise operation for Windows too.

File: pref/battery/battery.go
Comment (whereswaldon) [inline pref/battery/battery.go:33]:
```suggestion
// If the device doesn't rely on batteries it will be always true.
```

File: pref/battery/battery_windows.go
Comment (whereswaldon) [inline pref/battery/battery_windows.go:27]:
```suggestion
	if resp.BatteryLifePercent == _BatteryLifePercentUnknown || resp.BatteryFlag&_BatteryFlagNoSystemBattery > 0 {
```
I think this was an accidental over-generalization. The `BatteryLifePercent` value isn't a flag, so I think it's proper to do direct comparison on it.

File: pref/internal/xjni/jni_android.go
Comment (whereswaldon) [inline pref/internal/xjni/jni_android.go:1]:
```suggestion
// That package is used to reduce code duplication, when using JNI.
```

File: pref/internal/xjni/jni_android.go
Comment (whereswaldon) [inline pref/internal/xjni/jni_android.go:9]:
```suggestion
// DoInt invokes a static int method in the JVM and returns its results. lib is the path to the
// java class with the method, function is the name of the method, signature is the JNI
// string description of the method signature, and args allows providing parameters to
// the method.
func DoInt(lib string, function string, signature string, args ...jni.Value) (i int, err error) {
```

File: pref/internal/xjni/jni_android.go
Comment (whereswaldon) [inline pref/internal/xjni/jni_android.go:27]:
```suggestion
// DoString invokes a static String method in the JVM and returns its results. lib is the path to the
// java class with the method, function is the name of the method, signature is the JNI
// string description of the method signature, and args allows providing parameters to
// the method.
func DoString(lib string, function string, signature string, args ...jni.Value) (s string, err error) {
```

File: pref/locale/locale.go
Comment (whereswaldon) [inline pref/locale/locale.go:17]:
```suggestion
// Language is the preferred language of the end-user or the language of the operating system.
```

**Review by whereswaldon — CHANGES_REQUESTED** [review]
This is pretty much ready. I have some doc suggestions and one quick tweak to a windows API comparison, but I think I'm ready to merge once those items are taken care of.

**Comment by whereswaldon** [comment]
I see that you pushed, but I don't see all of my suggested changes addressed. Did you perhaps disagree with some or miss committing them?

**Review by whereswaldon — APPROVED** [review]
Okay! I think we're good to go. Merging now.

**Comment by whereswaldon** [comment]
Merged! https://git.sr.ht/~whereswaldon/gio-x/commit/0252dbbe449f43310beb5039d5a8880ff705ea0c

**Comment by Inkeliz** [comment]
The github mirror is outdated? 😶

**Comment by whereswaldon** [comment]
Yeah, looks like the job that mirrored it was stuck. I've fixed that and configured alerts for if this happens again.

**Pull request closed without merging**
